README.md
# Node Hugo

A basic cross-platform interface to the Hugo binary from Node.js that uses the 64-bit release binaries to support Linux, macOS, and Windows.

It should __just work™__

## System requirements

Runs on x64, arm, and arm64 on Linux, macOS, and Windows.

## Installation

```sh
npm i @small-tech/node-hugo
```

## Usage

```js
const Hugo = require('@small-tech/node-hugo')

(async function main () {

  const optionalPathToHugoConfiguration = 'optional/path/to/hugo/configuration/'

  const hugo = new Hugo()

  // Run a build.
  try {
    await hugo.build(optionalPathToHugoConfiguration)
  } catch (error) {
    console.log('There was an error building the site with Hugo'. error)
    return
  }

  console.log('Site build successful. Output:', hugo.output)

  // Start Hugo server. Returns a ChildProcess instance.
  const hugoServerProcess = hugo.createServer(optionalPathToHugoConfiguration)

  hugoServerProcess.on('error', (error) => {
    console.log('Hugo server encountered an error', error)
  })

  hugoServerProcess.stdout.on('data', (data) => {
    console.log(`[Hugo] ${data}`)
  })

  hugoServerProcess.stderr.on('data', (data) => {
    console.log(`[Hugo] [ERROR] ${data}`)
  })

  hugoServerProcess.on('close', (code) => {
    console.log('Hugo server process exited with code', code)
  })

  // Close the server after 5 seconds.
  setTimeout(() => {
    hugoServerProcess.kill()
  }, 5000)

})()
```

__(Version 1.2.0+)__ You can also run any Hugo command using the simple pass-through `command()` method:

```js
const Hugo = require('@small-tech/node-hugo')

(async function main () {

  const hugo = new Hugo()

  try {
    // Tell the built-in Hugo binary to create a new site at ./my-new-site/.
    await hugo.command('new site my-new-site')
  } catch (error) {
    console.log('There was an error creating the new site with Hugo'. error)
    return
  }

  console.log('Site build successful. Output:', hugo.output)
})()
```

__(Version 1.3.0+)__ For creating Hugo servers with an arbitrary argument string, you can use the `serverWithArgs()` method. This works like the blocking `command()` method but provides streaming output from the Hugo server.

__(Version 1.4.0+)__ You can retrieve the version of Hugo that’s bundled using the `version` property:

```js
const Hugo = require('@small-tech/node-hugo')
const hugo = new Hugo()
console.log(`Hugo version: ${hugo.version}`)
```

## Maintenance

### Update Hugo binaries

On Linux-esque operating systems (with Bash shell):

```sh
npm run update-hugo-bin
```

This will remove the current binaries, download the latest ones, and rename them properly.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2020-11-04

### Breaking change

__Upgrade Hugo to version 0.78.0 (November 3rd, 2020).__

(Previously bundled version was 0.64.1 from February 9th, 2020.)

There are breaking changes between these Hugo versions, so please read through the [Hugo release notes](https://github.com/gohugoio/hugo/releases).

We ran into the following two issues while upgrading our own sites:

  - In the `[outputs]` section of your _config.toml_ file, change `taxonomy` parameter to `tags` or `categories`. (See [this issue](https://source.small-tech.org/site.js/starters/starter-theme/-/issues/109).)

  - Change usage of parentheses in `if or` statements. (See [this commit](https://source.small-tech.org/site.js/starters/starter-theme/-/commit/2f116aa7f7f3cc4e8db028ad2d47b2ff1c4200f2).)

### Changed

  - Now includes separate arm64 binary.

## [1.5.0] - 2020-07-09

### Added

  - arm64 support.

## [1.4.1] - 2020-06-28

### Added

  - Option to disable build drafts for Hugo server; new fourth parameter in serve() method.

## [1.4.0] - 2020-05-05

### Added

  - `version` property that returns the Hugo version ([#3](https://source.small-tech.org/site.js/lib/node-hugo/-/issues/3)).

### Fixed

  - Incorrect require statements in code samples in readme ([#2](https://source.small-tech.org/site.js/lib/node-hugo/-/issues/2)).

## [1.3.0] - 2020-02-13

### Added

  - Generic hugo.serverWithArgs() method to run a Hugo server with any arbitrary argument string.

## [1.2.0] - 2020-02-12

### Added

  - Generic hugo.command() method to run any Hugo command by passing a regular Hugo command-line argument string.

## [1.1.1] - 2020-02-11

### Fixes

  - Fix regression due to version number change
  - Update tests

## [1.1.0] - 2020-02-11

### Changes

  - Upgrade Hugo to [version 0.64.1](https://github.com/gohugoio/hugo/releases/tag/v0.64.1).

## [1.0.1] - 2020-02-10

### Fixes

  - Disables fast render so Hugo updates are more consistent.

## [1.0.0] - 2020-02-07

Initial release.


test.js
const test = require('tape')
const fs = require('fs-extra')
const http = require('http')

const hugo = new (require('./index'))()

async function httpGet (url) {
  return new Promise((resolve, reject) => {
    http.get(url, (response) => {
      const statusCode = response.statusCode
      const location = response.headers.location

      // Reject if it’s not one of the status codes we are testing.
      if (statusCode !== 200 && statusCode !== 404 && statusCode !== 500 && statusCode !== 302) {
        reject({statusCode})
      }

      let body = ''
      response.on('data', _ => body += _)
      response.on('end', () => {
        resolve({statusCode, location, body})
      })
    })
  })
}


test('[hugo.command]', async t => {
  const sourcePath = 'test/a-new-site'

  if (fs.existsSync(sourcePath)) {
    fs.removeSync(sourcePath)
  }

  const args = `new site ${sourcePath}`

  let hasError = false
  let output = null
  try {
    // Tell the built-in Hugo binary to create a new site at ./my-new-site/.
    output = await hugo.command(args)
  } catch (error) {
    hasError = true
  }

  t.false(hasError, 'does not throw an error')
  t.true(output.includes('Congratulations! Your new Hugo site is created'), 'new site is generated')
  t.true(fs.existsSync(sourcePath), 'generated site exists where we expect it')

  const generatedFiles = fs.readdirSync(sourcePath)
  t.true(generatedFiles.join(',') === 'archetypes,config.toml,content,data,layouts,static,themes', 'contents of generated site are as expected')

  t.end()
})


test('[hugo.build] ', async t => {
  const sourcePath = 'test/site'
  const destinationPath = '../public' /* NB. relative to source path */
  const baseURL = 'http://localhost:1313'

  if (fs.existsSync(destinationPath)) {
    fs.removeSync(destinationPath)
  }

  let hasError = false
  let output = null
  try {
    output = await hugo.build(sourcePath, destinationPath, baseURL)
  } catch (error) {
    hasError = true
  }

  t.false(hasError, 'does not throw an error')

  const deflatedOutput = output.replace(/\s/g, '')

  // Ensure build statistics are as we expect.
  t.true(deflatedOutput.includes('Pages|2Paginatorpages|0Non-pagefiles|1Staticfiles|0Processedimages|0Aliases|0Sitemaps|1Cleaned|0'), 'console output statistics are as expected')

  const deflatedIndexPageHTML = fs.readFileSync('test/public/index.html', 'utf-8').replace(/\s/g, '')

  const expectedDeflatedIndexPageHTML = '<!DOCTYPEhtml><htmllang="en"><head><metaname="generator"content="Hugo0.78.0"/><metacharset="UTF-8"><metaname="viewport"content="width=device-width,initial-scale=1.0"><title>Yes!</title></head><body><h1>Yes!</h1><p>node-hugo<strong>works!</strong></p></body></html>'

  // Ensure generated HTML is as we expect.
  t.strictEquals(expectedDeflatedIndexPageHTML, deflatedIndexPageHTML, 'index page HTML is as expected')

  t.end()
})


test('[hugo.serve] ', async t => {
  const sourcePath = 'test/site'
  const destinationPath = '../public' /* NB. relative to source path */
  const baseURL = 'http://localhost:1313'

  if (fs.existsSync(destinationPath)) {
    fs.removeSync(destinationPath)
  }

  const {hugoServerProcess, hugoBuildOutput} = await hugo.serve(sourcePath, destinationPath, baseURL)
  const hugoBuildOutputDeflated = hugoBuildOutput.replace(/\s/g, '')

  t.false(hugoBuildOutputDeflated.includes('WARN'), 'build did not encounter a warning')
  t.true(hugoBuildOutputDeflated.includes('Pages|2'), 'two pages are rendered')
  t.true(hugoBuildOutputDeflated.includes('Non-pagefiles|1'), 'one non-page file is rendered')
  t.true(hugoBuildOutputDeflated.includes('Sitemaps|1'), 'one sitemap is rendered')

  hugoServerProcess.kill()

  await new Promise((resolve, reject) => {
    hugoServerProcess.on('exit', (code) => {
      t.assert(true, 'server process exited via kill()')
      resolve()
    })
  })

  t.end()
})


test('[hugo.version]', async t => {
  const output = await hugo.command('version')
  const versionFromHugo = output.match(/v(\d+\.\d+\.\d+)-/)[1]

  t.strictEquals(hugo.version, versionFromHugo, 'Hugo version is correct.')

  t.end()
})


index.js
////////////////////////////////////////////////////////////////////////////////
//
// @small-tech/node-hugo
//
// A basic cross-platform interface to the Hugo binary from Node.js that uses
// the 64-bit release binaries to support Linux, macOS, and Windows.
//
// Copyright ⓒ 2020 Aral Balkan. Licensed under AGPLv3 or later.
// Shared with ♥ by the Small Technology Foundation.
//
// Like this? Fund us!
// https://small-tech.org/fund-us
//
////////////////////////////////////////////////////////////////////////////////

const os = require('os')
const path = require('path')
const fs = require('fs-extra')

const util = require('util')
const childProcess = require('child_process')

const exec = util.promisify(childProcess.exec)
const spawn = childProcess.spawn

const homeDir = os.homedir()

const hugoVersion = '0.78.0'

class HugoError extends Error {
  constructor (message, output) {
    super(message)
    this.sdterr = output
    this.name = 'HugoError'
  }
}

class Hugo {
  constructor (nodeHugoDir = path.join(homeDir, '.small-tech.org', 'node-hugo')) {

    this.machine = {
      platform: os.platform(),
      architecture: os.arch()
    }

    this.nodeHugoDir = nodeHugoDir

    // Ensure the node-hugo directory exists.
    fs.ensureDirSync(nodeHugoDir)

    this.hugoBinaryPath = this.hugoBinaryForThisMachine()
  }

  //
  // Public.
  //

  // Returns the Hugo version.
  get version () {
    return hugoVersion
  }

  // Runs a generic, blocking Hugo command using the passed arguments.
  async command (args) {
    const hugoCommand = `${this.hugoBinaryPath} ${args}`
    const options = {
      env: process.env,
      stdio: 'pipe',
    }
    const result = await exec(hugoCommand, options)

    if (result.stderr !== '') {
      // There was an error.
      throw new HugoError(`Hugo command failed (arguments: ${args})`, result.stderr)
    }

    return result.stdout
  }

  // Builds the Hugo source from sourcePath and writes the output at
  // destinationPath. Note that destinationPath is relevant to sourcePath
  // (don’t shoot me, this is a Hugo convention so I’m mirroring it for
  // consistency with regular Hugo usage). The returned result is the
  // object returned from the exec() call with stdout and stderr properties.
  async build (sourcePath = '.', destinationPath = 'public/', baseURL = 'http://localhost:1313') {
    const hugoBuildCommand = `--source=${sourcePath} --destination=${destinationPath} --baseURL=${baseURL}`
    return await this.command(hugoBuildCommand)
  }

  // Starts a Hugo server with defaults set for rendering to disk and using external live reload.
  // (These are the defaults for our current use case on Site.js).
  //
  // Returns an object that contains a promise that’s resolved to an object with the following
  // properties once the initial Hugo build is complete:
  //
  //   - a reference to the Hugo server process
  //   - the Hugo output so far
  //
  serve (sourcePath = '.', destinationPath = 'public/', baseURL = 'http://localhost:1313', buildDrafts = true) {
    const args = [
      'server',
      `--source=${sourcePath}`,
      `--destination=${destinationPath}`,
      `--baseURL=${baseURL}`,
      '--renderToDisk',
      '--disableLiveReload',
      '--appendPort=false',
      '--disableFastRender'
    ]

    if (buildDrafts) {
      args.push('--buildDrafts')
    }

    return this.serverWithArgs(args)
  }

  // Starts a generic Hugo server.
  serverWithArgs (args) {
    // Args should be an array. Automatically convert an arguments string to one.
    if (typeof args === 'string') {
      args = args.split(' ')
    }
    const options = { env: process.env }
    const hugoServerProcess = spawn(this.hugoBinaryPath, args, options)

    const hugoServerPromise = new Promise((resolve, reject) => {
      // Wait for the line telling us that the build is complete and that
      // the server is ready and only then return.

      let hugoBuildOutput = ''
      let buildComplete = false

      const stdoutHandler = (data) => {
        const lines = data.toString('utf-8').split('\n')

        lines.forEach(line => {
          hugoBuildOutput += `\n${line}`

          if (line.startsWith('Built in')) {
            // The site has been built. Let’s resolve the promise.
            buildComplete = true
          }
        })

        if (buildComplete) {
          // Clean up. If the consumer wants to, they can add their own to the
          // hugo server process instance that we will return.
          hugoServerProcess.stdout.removeAllListeners()
          hugoServerProcess.stderr.removeAllListeners()

          // OK.
          resolve({
            hugoServerProcess,
            hugoBuildOutput
          })
        }
      }

      const stderrHandler = (data) => {
        const errorMessage = data.toString('utf-8')
        reject(errorMessage)
      }

      hugoServerProcess.stdout.on('data', stdoutHandler)
      hugoServerProcess.stderr.on('data', stderrHandler)
    })

    return hugoServerPromise
  }

  //
  // Private.
  //

  // Returns the Hugo binary for this machine (platform + architecture) and
  // throws an error if there isn’t one for it.
  //
  // Note: this expects the Hugo binaries to be manually renamed prior to being
  // ===== added to the hugo-bin folder. The naming convention we use is the same
  //       as the one we use in the Auto Encrypt Localhost project for the mkcert binaries.
  hugoBinaryForThisMachine () {
    const platformMap = {
      linux: 'linux',
      darwin: 'darwin',
      win32: 'windows'
    }

    const architectureMap = {
      arm: 'arm',
      arm64: 'arm64',
      x64: 'amd64'
    }

    const platform = platformMap[this.machine.platform]
    const architecture = architectureMap[this.machine.architecture]

    if (platform === undefined) throw new Error('Unsupported platform', this.machine.platform)
    if (architecture === undefined) throw new Error('Unsupported architecture', this.machine.architecture)

    const hugoBinaryName = `hugo-v${hugoVersion}-${platform}-${architecture}`
    const hugoBinaryRelativePath = path.join('hugo-bin', hugoBinaryName)
    let hugoBinaryInternalPath = path.join(__dirname, hugoBinaryRelativePath)

    if (platform === 'windows') hugoBinaryInternalPath += '.exe'

    // Check if the platform + architecture combination is supported.
    if (!fs.existsSync(hugoBinaryInternalPath)) throw new Error(`[node-hugo] Unsupported platform + architecture combination for ${platform}-${architecture}`)

    // Copy the Hugo binary to the external node-hugo directory so that we can call execSync() on it if
    // the app using this module is wrapped into an executable using Nexe (https://github.com/nexe/nexe) – like
    // Site.js (https://sitejs.org) is, for example. We use readFileSync() and writeFileSync() as
    // Nexe does not support copyFileSync() yet (see https://github.com/nexe/nexe/issues/607).
    const hugoBinaryExternalPath = path.join(this.nodeHugoDir, hugoBinaryName)

    if (!fs.existsSync(hugoBinaryExternalPath)) {
      try {
        const hugoBinaryBuffer = fs.readFileSync(hugoBinaryInternalPath, 'binary')
        fs.writeFileSync(hugoBinaryExternalPath, hugoBinaryBuffer, {encoding: 'binary', mode: 0o755})
      } catch (error) {
        throw new Error(` 🤯 [node-hugo] Panic: Could not copy Hugo binary to external directory: ${error.message}`)
      }
    }

    return hugoBinaryExternalPath
  }
}

module.exports = Hugo
